### ISSUE TITLE
How to build handtracking CPU AAR or postracking CPU AAR to reduce dependence on GPU？

### ISSUE THREAD

**Issue by DandyLorenz** [google/mediapipe#1994]
In order to reduce the performance of mediapipe in low-end devices, or how to reduce the requirement of mediapipe on hardware devices (CPU or GPU), I plan to **use CPU** to run **hand tracking** or **post tracking**. 
My build environment has no problem and can successfully compile and run **Android AAR**. I need to know whether mediapipe supports hand or pose tracking CPU ? The following is the modification of my build pose CPU AAR (it will report an error when it is used in the end):
vi **BUILD** file:
`load("//mediapipe/java/com/google/mediapipe:mediapipe_aar.bzl", "mediapipe_aar")
mediapipe_aar(
    name = "my_posetrackingcpu_aar",
    calculators = ["//mediapipe/graphs/pose_tracking:pose_tracking_cpu_deps",
],
)`

**Build Command:**
`bazel build -c opt --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --fat_apk_cpu=arm64-v8a,armeabi-v7a \
mediapipe/examples/android/src/java/com/google/mediapipe/apps/my_posetrackinggpu_aar:my_posetrackinggpu_aar --verbose_failures`

I build and use **pose_tracking_cpu.binarypb**
`bazel build -c opt mediapipe/graphs/pose_tracking:pose_tracking_cpu_binary_graph`

I will add the result and screenshot of the error in the comments later.Urgent need your help, thank you

**Comment by DandyLorenz** [comment]
The parameters of my hardware are as follows:
CPU model: rk30board
CPU cores: 6
CPU architecture: arm-64-v8a, armeabi-v7a, armeabi
Maximum CPU frequency: 1.42ghz X4; 1.80Ghz x 2
SDK version: 25
Build: rk3399_ all_ userdebug7.1.2
System memory: 3.76gb
Available memory: 2.85gb
Graphics processor: mali-t860
GPU:ARM
OpenGL: OpenGL ES CM 1.1v1.r18p0
Running the GPU version of handtracking GPU on the device, it takes about 30 seconds from the start of the camera to the appearance of the image recognition gesture, and it takes about 30 seconds to wait on the black screen. It's a bit stuck when it runs. After a period of time, it will stop

**Comment by mcclanahoochie** [comment]
Hi, yes, any example MediaPipe provides can be made to work on CPU/GPU, by modifying the graph and build flags.

Typically (in most, but not all graphs), you will see graph stream names with IMAGE_GPU to denote GPU data, or just IMAGE for CPU.

For all Android examples, it is assumed that the input and output streams are on GPU (by default), unless you have a custom MainActivity.java that can send/out in cpu frames. 

Any logs you can provide can allow more help here.

**Comment by DandyLorenz** [comment]
yes，I try to modify some BUILD file About its calculators content，such as,I build handtracking aar for android,I Modify as follows：

**Comment by DandyLorenz** [comment]
yes，I try to modify some BUILD file About its calculators content，such as,I build handtracking aar for android,I Modify as follows：

**Comment by DandyLorenz** [comment]
yes，I try to modify some BUILD file About its calculators content，such as,I build posetracking aar for android,I Modify as follows：
1.Create folder my_posetrackingcpu_aar and create file BUILD below:
# vi BUILD
`load("//mediapipe/java/com/google/mediapipe:mediapipe_aar.bzl", "mediapipe_aar")

mediapipe_aar(
  name = "my_posetrackingcpu_aar",
  calculators = [
    "//mediapipe/graphs/pose_tracking:pose_tracking_cpu_deps",
  ],

)`
2.run linux  command for build aar for android:
`bazel build -c opt --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --fat_apk_cpu=arm64-v8a,armeabi-v7a \ mediapipe/examples/android/src/java/com/google/mediapipe/apps/my_posetrackingcpu_aar:my_posetrackingcpu_aar --verbose_failures`
3.build graphs command:
`bazel build -c opt mediapipe/graphs/pose_tracking:pose_tracking_cpu_binary_graph`
4. I have a custom MainActivity.java,I don't think it's different from the GPU version.this is simple:
``` 
   public class MainActivity extends AppCompatActivity {

    private static final String TAG = "pose_tracking";
    static
    {
        System.loadLibrary("mediapipe_jni");
        try
        {
            System.loadLibrary("opencv_java3");
        }
        catch (UnsatisfiedLinkError localUnsatisfiedLinkError)
        {
            Log.e(TAG,"System.loa   dLibrary opencv_java3 error,will try opencv_java4");
            System.loadLibrary("opencv_java4");
        }
    }
    private SurfaceTexture previewFrameTexture = null;
    private SurfaceView previewDisplayView = null;
    private CameraXPreviewHelper cameraHelper;
    private static final CameraHelper.CameraFacing CAMERA_FACING = CameraHelper.CameraFacing.FRONT;
    private EglManager eglManager = null;
    private ExternalTextureConverter converter;
    private static final boolean FLIP_FRAMES_VERTICALLY = true;
    private FrameProcessor processor;
    
    private static final String BINARY_GRAPH_NAME = "pose_tracking_cpu.binarypb";
    private static final String INPUT_VIDEO_STREAM_NAME = "input_video";
    private static final String OUTPUT_VIDEO_STREAM_NAME = "output_video";
    

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.w("myInfo","onCreate");
        setContentView(R.layout.activity_main);
        try {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
           
            AndroidAssetUtil.initializeNativeAssetManager(this);
            previewDisplayView = new SurfaceView(this);
            setupPreviewDisplayView();

            PermissionHelper.checkAndRequestCameraPermissions(this);
            
            if(eglManager == null){
                eglManager = new EglManager(null);
                Log.w("myInfo","new EglManager()");
            }

            processor = new FrameProcessor(
                    this,
                    eglManager.getNativeContext(),
                    BINARY_GRAPH_NAME,
                    INPUT_VIDEO_STREAM_NAME,
                    OUTPUT_VIDEO_STREAM_NAME);
            processor.getVideoSurfaceOutput().setFlipY(FLIP_FRAMES_VERTICALLY);

            HashMap< String, PacketCallback> packetCallbacks = new HashMap< String, PacketCallback>();
            packetCallbacks.put("pose_landmarks", (packet -> {
                String temp = this.getNormalizedLandmarkListString(2, packet);
            }));
            for (Map.Entry<String, PacketCallback> pair: packetCallbacks.entrySet()) {
                processor.addPacketCallback(pair.getKey(), pair.getValue());
            }

        } catch (Exception paramBundle) {
            for (;;)
            {
                Log.e("MainActivity", "***********Cannot find application info: " + paramBundle);
            }
        }

    }

    protected  String getNormalizedLandmarkListString(int listIndex, Packet packet) {
        LandmarkProto.NormalizedLandmarkList landmarkList = this.getNormalizedLandmarkList(packet);
      
        if (landmarkList == null || landmarkList.getLandmarkCount() <= 0){
            Log.w(TAG, "landmarks("+listIndex+"): " + "landmarkList == null || landmarkList.getLandmarkCount() <= 0");
        }
        String result = "";
        int landmarkIndex = 0;
        for (LandmarkProto.NormalizedLandmark landmark : landmarkList.getLandmarkList()) {
            String temp = listIndex + "," + landmarkIndex + "," + new BigDecimal(landmark.getX()).toPlainString()
                    + "," + new BigDecimal(landmark.getY()).toPlainString() + "," + new BigDecimal(landmark.getZ()).toPlainString();
            if (landmarkIndex >= 1) {
                result += ",";
            }
            result += "[" + temp + "]";
            ++landmarkIndex;
        }
        return "[-1," + packet.getTimestamp() + ",0,0,0]," + result;
    }

    private LandmarkProto.NormalizedLandmarkList getNormalizedLandmarkList(Packet packet) {
        byte[] landmarksRaw = PacketGetter.getProtoBytes(packet);
        LandmarkProto.NormalizedLandmarkList landmarks;
        try {
            landmarks = LandmarkProto.NormalizedLandmarkList.parseFrom(landmarksRaw);
        } catch (InvalidProtocolBufferException e) {
            Log.d(TAG, e.getMessage()); //            e.printStackTrace();
            return null;
        }
        if (landmarks.getLandmarkCount() <= 0){
            return null;
        }
        return landmarks;
    }

    private void setupPreviewDisplayView() {
        previewDisplayView.setVisibility(View.GONE);
        ViewGroup viewGroup = findViewById(R.id.preview_display_layout);
        viewGroup.addView(previewDisplayView);
        previewDisplayView
                .getHolder()
                .addCallback(
                        new SurfaceHolder.Callback() {
                            @Override
                            public void surfaceCreated(SurfaceHolder holder) {
                                processor.getVideoSurfaceOutput().setSurface(holder.getSurface());
                            }

                            @Override
                            public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
                                Size viewSize = new Size(width, height);
                                Size displaySize = cameraHelper.computeDisplaySizeFromViewSize(viewSize);
                                boolean isCameraRotated = cameraHelper.isCameraRotated();
                                Size size2 = cameraHelper.getFrameSize();
                                Log.w(TAG,"after getFrameSize getWidth:" + size2.getWidth() + " getHeight: " + size2.getHeight() );
                                if(converter == null){
                                    Log.w(TAG,"****** converter == null");
                                }
                                converter.setSurfaceTextureAndAttachToGLContext(previewFrameTexture,
                                        isCameraRotated ? displaySize.getHeight() : displaySize.getWidth(),
                                        isCameraRotated ? displaySize.getWidth() : displaySize.getHeight()
                                );
                                try {
                                    previewFrameTexture.updateTexImage();
                                }catch (Exception e){
                                    Log.w("myInfo","updateTexImage Exception:" + e.getMessage());
                                }
                            }

                            @Override
                            public void surfaceDestroyed(SurfaceHolder holder) {
                                processor.getVideoSurfaceOutput().setSurface(null);
                            }
                        });
    }

    @Override
    public void onRequestPermissionsResult(
            int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        PermissionHelper.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onPause() {
        super.onPause();
        converter.close();
        converter = null;
        previewDisplayView.setVisibility(View.GONE);
        previewFrameTexture.release();
       
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (eglManager == null){
            Log.w("myInfo","***** eglManager == null");
        }
        converter = new ExternalTextureConverter(eglManager.getContext());
        converter.setFlipY(FLIP_FRAMES_VERTICALLY);
        converter.setConsumer(processor);

        if (PermissionHelper.cameraPermissionsGranted(this)) {
            startCamera();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    public void startCamera() {
        cameraHelper = new CameraXPreviewHelper();
        cameraHelper.setOnCameraStartedListener(
                surfaceTexture -> {
                    Log.w("myInfo","OnCameraStarted");
                    if(surfaceTexture == null){
                        Log.w("myInfo","***** surfaceTexture == null");
                    } else {
                        surfaceTexture.setOnFrameAvailableListener((SurfaceTexture var1) ->{
                            Log.w("myInfo","Frame available");
                        });
                    }
                    previewFrameTexture = surfaceTexture;
                    previewDisplayView.setVisibility(View.VISIBLE);
                });
        cameraHelper.startCamera(this, CAMERA_FACING, /*surfaceTexture=*/ null);
    }
}```
5.Running on a real phone,show error logs：

![clipboard](https://user-images.githubusercontent.com/40449226/120957996-51b20400-c789-11eb-9007-9ffb892fa3be.png)
`com.google.mediapipe.framework.MediaPipeException: not found: ValidatedGraphConfig Initialization failed.
    No registered object with name: PoseLandmarkGpu; Unable to find Calculator "PoseLandmarkGpu"
    No registered object with name: PoseRendererGpu; Unable to find Calculator "PoseRendererGpu"
`
6.My question：
1.Which step did I do wrong or need to be improved?
2.Is there a detailed build tutorial such  as pose trackng cpu or hand tracking cpu?

**Comment by DandyLorenz** [comment]
yes，I try to modify some BUILD file About its calculators content，such as,I build posetracking aar for android,I Modify as follows：
1.Create folder my_posetrackingcpu_aar and create file BUILD below:
`vi BUILD`
`load("//mediapipe/java/com/google/mediapipe:mediapipe_aar.bzl", "mediapipe_aar")

mediapipe_aar(
name = "my_posetrackingcpu_aar",
calculators = [
"//mediapipe/graphs/pose_tracking:pose_tracking_cpu_deps",
],
)
2.run linux command for build aar for android:
`bazel build -c opt --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --fat_apk_cpu=arm64-v8a,armeabi-v7a \ mediapipe/examples/android/src/java/com/google/mediapipe/apps/my_posetrackingcpu_aar:my_posetrackingcpu_aar --verbose_failures`
3.build graphs command:
`bazel build -c opt mediapipe/graphs/pose_tracking:pose_tracking_cpu_binary_graph`
4. I have a custom MainActivity.java,I don't think it's different from the GPU version.this is simple:

   public class MainActivity extends AppCompatActivity {

    private static final String TAG = "pose_tracking";
    static
    {
        System.loadLibrary("mediapipe_jni");
        try
        {
            System.loadLibrary("opencv_java3");
        }
        catch (UnsatisfiedLinkError localUnsatisfiedLinkError)
        {
            Log.e(TAG,"System.loa   dLibrary opencv_java3 error,will try opencv_java4");
            System.loadLibrary("opencv_java4");
        }
    }
    private SurfaceTexture previewFrameTexture = null;
    private SurfaceView previewDisplayView = null;
    private CameraXPreviewHelper cameraHelper;
    private static final CameraHelper.CameraFacing CAMERA_FACING = CameraHelper.CameraFacing.FRONT;
    private EglManager eglManager = null;
    private ExternalTextureConverter converter;
    private static final boolean FLIP_FRAMES_VERTICALLY = true;
    private FrameProcessor processor;
    
    private static final String BINARY_GRAPH_NAME = "pose_tracking_cpu.binarypb";
    private static final String INPUT_VIDEO_STREAM_NAME = "input_video";
    private static final String OUTPUT_VIDEO_STREAM_NAME = "output_video";
    

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.w("myInfo","onCreate");
        setContentView(R.layout.activity_main);
        try {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
           
            AndroidAssetUtil.initializeNativeAssetManager(this);
            previewDisplayView = new SurfaceView(this);
            setupPreviewDisplayView();

            PermissionHelper.checkAndRequestCameraPermissions(this);
            
            if(eglManager == null){
                eglManager = new EglManager(null);
                Log.w("myInfo","new EglManager()");
            }

            processor = new FrameProcessor(
                    this,
                    eglManager.getNativeContext(),
                    BINARY_GRAPH_NAME,
                    INPUT_VIDEO_STREAM_NAME,
                    OUTPUT_VIDEO_STREAM_NAME);
            processor.getVideoSurfaceOutput().setFlipY(FLIP_FRAMES_VERTICALLY);

            HashMap< String, PacketCallback> packetCallbacks = new HashMap< String, PacketCallback>();
            packetCallbacks.put("pose_landmarks", (packet -> {
                String temp = this.getNormalizedLandmarkListString(2, packet);
            }));
            for (Map.Entry<String, PacketCallback> pair: packetCallbacks.entrySet()) {
                processor.addPacketCallback(pair.getKey(), pair.getValue());
            }

        } catch (Exception paramBundle) {
            for (;;)
            {
                Log.e("MainActivity", "***********Cannot find application info: " + paramBundle);
            }
        }

    }

    protected  String getNormalizedLandmarkListString(int listIndex, Packet packet) {
        LandmarkProto.NormalizedLandmarkList landmarkList = this.getNormalizedLandmarkList(packet);
      
        if (landmarkList == null || landmarkList.getLandmarkCount() <= 0){
            Log.w(TAG, "landmarks("+listIndex+"): " + "landmarkList == null || landmarkList.getLandmarkCount() <= 0");
        }
        String result = "";
        int landmarkIndex = 0;
        for (LandmarkProto.NormalizedLandmark landmark : landmarkList.getLandmarkList()) {
            String temp = listIndex + "," + landmarkIndex + "," + new BigDecimal(landmark.getX()).toPlainString()
                    + "," + new BigDecimal(landmark.getY()).toPlainString() + "," + new BigDecimal(landmark.getZ()).toPlainString();
            if (landmarkIndex >= 1) {
                result += ",";
            }
            result += "[" + temp + "]";
            ++landmarkIndex;
        }
        return "[-1," + packet.getTimestamp() + ",0,0,0]," + result;
    }

    private LandmarkProto.NormalizedLandmarkList getNormalizedLandmarkList(Packet packet) {
        byte[] landmarksRaw = PacketGetter.getProtoBytes(packet);
        LandmarkProto.NormalizedLandmarkList landmarks;
        try {
            landmarks = LandmarkProto.NormalizedLandmarkList.parseFrom(landmarksRaw);
        } catch (InvalidProtocolBufferException e) {
            Log.d(TAG, e.getMessage()); //            e.printStackTrace();
            return null;
        }
        if (landmarks.getLandmarkCount() <= 0){
            return null;
        }
        return landmarks;
    }

    private void setupPreviewDisplayView() {
        previewDisplayView.setVisibility(View.GONE);
        ViewGroup viewGroup = findViewById(R.id.preview_display_layout);
        viewGroup.addView(previewDisplayView);
        previewDisplayView
                .getHolder()
                .addCallback(
                        new SurfaceHolder.Callback() {
                            @Override
                            public void surfaceCreated(SurfaceHolder holder) {
                                processor.getVideoSurfaceOutput().setSurface(holder.getSurface());
                            }

                            @Override
                            public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
                                Size viewSize = new Size(width, height);
                                Size displaySize = cameraHelper.computeDisplaySizeFromViewSize(viewSize);
                                boolean isCameraRotated = cameraHelper.isCameraRotated();
                                Size size2 = cameraHelper.getFrameSize();
                                Log.w(TAG,"after getFrameSize getWidth:" + size2.getWidth() + " getHeight: " + size2.getHeight() );
                                if(converter == null){
                                    Log.w(TAG,"****** converter == null");
                                }
                                converter.setSurfaceTextureAndAttachToGLContext(previewFrameTexture,
                                        isCameraRotated ? displaySize.getHeight() : displaySize.getWidth(),
                                        isCameraRotated ? displaySize.getWidth() : displaySize.getHeight()
                                );
                                try {
                                    previewFrameTexture.updateTexImage();
                                }catch (Exception e){
                                    Log.w("myInfo","updateTexImage Exception:" + e.getMessage());
                                }
                            }

                            @Override
                            public void surfaceDestroyed(SurfaceHolder holder) {
                                processor.getVideoSurfaceOutput().setSurface(null);
                            }
                        });
    }

    @Override
    public void onRequestPermissionsResult(
            int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        PermissionHelper.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onPause() {
        super.onPause();
        converter.close();
        converter = null;
        previewDisplayView.setVisibility(View.GONE);
        previewFrameTexture.release();
       
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (eglManager == null){
            Log.w("myInfo","***** eglManager == null");
        }
        converter = new ExternalTextureConverter(eglManager.getContext());
        converter.setFlipY(FLIP_FRAMES_VERTICALLY);
        converter.setConsumer(processor);

        if (PermissionHelper.cameraPermissionsGranted(this)) {
            startCamera();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    public void startCamera() {
        cameraHelper = new CameraXPreviewHelper();
        cameraHelper.setOnCameraStartedListener(
                surfaceTexture -> {
                    Log.w("myInfo","OnCameraStarted");
                    if(surfaceTexture == null){
                        Log.w("myInfo","***** surfaceTexture == null");
                    } else {
                        surfaceTexture.setOnFrameAvailableListener((SurfaceTexture var1) ->{
                            Log.w("myInfo","Frame available");
                        });
                    }
                    previewFrameTexture = surfaceTexture;
                    previewDisplayView.setVisibility(View.VISIBLE);
                });
        cameraHelper.startCamera(this, CAMERA_FACING, /*surfaceTexture=*/ null);
    }
}```
5.Running on a real phone,show error logs：

![clipboard](https://user-images.githubusercontent.com/40449226/120957996-51b20400-c789-11eb-9007-9ffb892fa3be.png)
`com.google.mediapipe.framework.MediaPipeException: not found: ValidatedGraphConfig Initialization failed.
    No registered object with name: PoseLandmarkGpu; Unable to find Calculator "PoseLandmarkGpu"
    No registered object with name: PoseRendererGpu; Unable to find Calculator "PoseRendererGpu"
`
6.My question：
1.Which step did I do wrong or need to be improved?
2.Is there a detailed build tutorial such  as pose trackng cpu or hand tracking cpu?

**Comment by sgowroji** [comment]
Hi @DandyLorenz, Could you please confirm the above is duplicate of this https://github.com/google/mediapipe/issues/2126

**Comment by DandyLorenz** [comment]
No, it isn't.The reasons are: 
1. I basically did not change the graph file
2. I am building aar for Android，But it is to build a python program.

**Comment by mcclanahoochie** [comment]
> Unable to find Calculator "PoseRendererGpu"

the GPU graph is being linked somehow, can you double check you are compiling/linking the CPU version? 

even if you are, the MainActivity.java you post here will be trying to send/receive GPU frames in/out of the graph. You will need GpuBufferToImageFrameCalculator at the beginning of the CPU graph and ImageFrameToGpuBufferCalculator at the end of the CPU graph, see https://github.com/google/mediapipe/blob/master/mediapipe/graphs/face_detection/face_detection_mobile_cpu.pbtxt#L37 for a "CPU" android graph.
